### PR TITLE
fix reward actor to use miner owner

### PR DESCRIPTION
**Pull request by whyrusleeping**

### CODE REVIEW

**Comment by anorth** [comment]
This is explicitly not what I want it to do. Maybe the slack conversation wasn't conclusive.

**Comment by anorth** [comment]
To be more clear: I think a better solution is that the reward actor holds balances on account for the miner actor (the same way balances are held in the market and power actors). Then, upon withdrawal only, the miner's owner address is looked up by calling the MinerControlAddrs method and the payment made to that address.

**Comment by anorth** [comment]
Ref: https://github.com/filecoin-project/specs/issues/787

**Comment by whyrusleeping** [comment]
That assumes that the reward actor is only ever used to pay out to miner actors, which is probably a fine assumption to make. but its definitely additional assumptions (at one weird point in time, multisig vesting also happened in here)

**Comment by whyrusleeping** [comment]
@anorth Addressed that, should be good to go

File: actors/builtin/reward/reward_actor.go [hunk 1]
```diff
@@ -49,10 +49,9 @@ func (a Actor) Constructor(rt vmr.Runtime, _ *adt.EmptyValue) *adt.EmptyValue {
 }
 
 type AwardBlockRewardParams struct {
-	Miner        addr.Address
-	Penalty      abi.TokenAmount // penalty for including bad messages in a block
-	GasReward    abi.TokenAmount // gas reward from all gas fees in a block
-	NominalPower abi.StoragePower
+	Miner     address.Address
+	Penalty   abi.TokenAmount // penalty for including bad messages in a block
+	GasReward abi.TokenAmount // gas reward from all gas fees in a block
```
Comment (anorth) [inline actors/builtin/reward/reward_actor.go]: Just FYI, I believe we'll need to add a winning ticket count parameter here later.

File: actors/builtin/reward/reward_actor.go [hunk 2]
```diff
@@ -67,10 +66,14 @@ type AwardBlockRewardParams struct {
...
 	rt.ValidateImmediateCallerIs(builtin.SystemActorAddr)
-	AssertMsg(params.Miner.Protocol() == addr.ID,
-		"miner address must be ID-address protocol (%d), was protocol %d", addr.ID, params.Miner.Protocol())
 	AssertMsg(rt.CurrentBalance().GreaterThanEqual(params.GasReward),
 		"actor current balance %v insufficient to pay gas reward %v", rt.CurrentBalance(), params.GasReward)
+
+	miner, ok := rt.ResolveAddress(params.Miner)
+	if !ok {
+		rt.Abortf(exitcode.ErrIllegalState, "failed to resolve given owner address")
+	}
```
Comment (anorth) [inline actors/builtin/reward/reward_actor.go]: The sender here will be the VM/interpreter, which has no excuse not to provide an ID address (which it gets from the block header). So I prefer the assertion, but this is ok.

File: actors/builtin/reward/reward_actor.go [hunk 3]
```diff
@@ -109,14 +112,20 @@ func (a Actor) AwardBlockReward(rt vmr.Runtime, params *AwardBlockRewardParams)
 	return &adt.EmptyValue{}
 }
 
-func (a Actor) WithdrawReward(rt vmr.Runtime, _ *adt.EmptyValue) *adt.EmptyValue {
+func (a Actor) WithdrawReward(rt vmr.Runtime, minerin *address.Address) *adt.EmptyValue {
 	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
```
Comment (anorth) [inline actors/builtin/reward/reward_actor.go]: I concur that the miner address needs to be a parameter, but in that case we probably want to lock down the caller to be the owner (or possibly the owner|worker). Wide open means _anyone_ could cause a miner's reward to be paid out to the owner account, which it's hard to see a legitimate reason for, and could be part of an attempted theft.

So I suggest `ValidateImmediateAddr(owner)` after resolving the owner.

File: actors/builtin/reward/reward_actor.go (hunk 1 above)
Comment (whyrusleeping) [reply actors/builtin/reward/reward_actor.go]: Ah, yeah. we will need that.

File: actors/builtin/reward/reward_actor.go (hunk 3 above)
Comment (whyrusleeping) [reply actors/builtin/reward/reward_actor.go]: I can see many reasons to not want to use your owner key here, I think restricting to owner or worker makes sense

**Pull request merged**